File: python/src/aoc_2022_15.py
"""Advent of Code 2022 - Day 15."""


from collections.abc import Iterable
from doctest import testmod
from itertools import combinations
from re import fullmatch
from sys import stdin

from coord import Coord, manhattan


class Zone:
    """Main object for day 15."""

    def __init__(self, lines: Iterable[str]) -> None:
        self.sensors: list[Coord] = []
        self.closest_beacons: list[Coord] = []
        x_minima: list[int] = []
        x_maxima: list[int] = []
        for line in lines:
            m = fullmatch(
                r"Sensor at x=(-?\d+), y=(-?\d+): closest beacon is at x=(-?\d+), y=(-?\d+)",
                line.strip(),
            )
            assert m is not None
            s = Coord(int(m.group(1)), int(m.group(2)))
            c = Coord(int(m.group(3)), int(m.group(4)))
            self.sensors.append(s)
            self.closest_beacons.append(c)
            d = manhattan(s, c)
            x_minima.append((s + Coord(-d, 0)).x)
            x_maxima.append((s + Coord(d, 0)).x)
        self.n = len(self.sensors)
        self.x_min = min(x_minima)
        self.x_max = max(x_maxima)

    def must_be_empty(self, x: int, y: int) -> bool:
        """Test if the given cell must be empty."""
        test_beacon = Coord(x, y)
        # print("Testing", test_beacon)
        for sensor, closest_beacon in zip(self.sensors, self.closest_beacons):
            if test_beacon == closest_beacon:
                break
            closest_beacon_distance = manhattan(sensor, closest_beacon)
            test_distance = manhattan(sensor, test_beacon)
            if test_distance <= closest_beacon_distance:
                # print("Forbidden", sensor)
                return True
        return False

    def empty_in_row(self, y: int) -> int:
        """Return number of positions with given y coordinate that must be empty.

        >>> z = Zone(TEST_DATA)
        >>> z.empty_in_row(10)
        26
        """
        count = 0
        for x in range(self.x_min, self.x_max + 1):
            count += self.must_be_empty(x, y)
        return count

    def find_beacon(self) -> Coord:
        """Find the only possible beacon location.

        Uses the fact that this unique location must be just outside the
        forbidden radius of two beacons.

        >>> Zone(TEST_DATA).find_beacon()
        Coord(x=14, y=11)
        """
        sensor_radius: dict[Coord, int] = {
            s: manhattan(s, b) for s, b in zip(self.sensors, self.closest_beacons)
        }
        x_plus_y_constraints: set[int] = set()
        x_minus_y_constraints: set[int] = set()
        for s1, s2 in combinations(self.sensors, 2):
            d = manhattan(s1, s2)
            r1 = sensor_radius[s1]
            r2 = sensor_radius[s2]
            if d == (r1 + 1) + (r2 + 1):
                constraint_sign, constraint_value = get_constraint(s1, s2, r1, r2)
                if constraint_sign:
                    x_plus_y_constraints.add(constraint_value)
                else:
                    x_minus_y_constraints.add(constraint_value)
        assert len(x_plus_y_constraints) > 0 and len(x_minus_y_constraints) > 0
        solutions: set[Coord] = set()
        for p_con in x_plus_y_constraints:
            for m_con in x_minus_y_constraints:
                if p_con + m_con % 2 == 1 or p_con - m_con % 2 == 1:
                    continue
                x = (p_con + m_con) // 2
                y = (p_con - m_con) // 2
                solutions.add(Coord(x, y))
        for solution in solutions:
            if all(
                manhattan(solution, sensor) > sensor_radius[sensor]
                for sensor in self.sensors
            ):
                return solution
        raise ValueError("No valid solution found")


def get_constraint(s1: Coord, s2: Coord, r1: int, r2: int) -> tuple[bool, int]:
    """Return the constraint on x+y or x-y.

    Points must have distance r1+1 from s1 and distance r2+1 from s2.
    Returns (True, z) for x+y=z and (False, z) for x-y=z.
    """
    # Ignore degenerate cases
    assert s1.x != s2.x and s1.y != s2.y
    # Wlg, flip so s1.x is less than s2.x
    if s1.x > s2.x:
        s1, s2 = s2, s1
        r1, r2 = r2, r1
    assert s1.x < s2.x
    if s1.y < s2.y:
        assert s1.x + s1.y + r1 + 1 == s2.x + s2.y - r2 - 1
        return (True, s1.x + s1.y + r1 + 1)
        # print(f"x+y = {s1.x+s1.y+r1+1} = {s2.x+s2.y-r2-1}")
    assert s1.x - s1.y + r1 + 1 == s2.x - s2.y - r2 - 1
    return (False, s1.x - s1.y + r1 + 1)


def encode(c: Coord) -> int:
    """Encode the coordinate."""
    return c.x * 4_000_000 + c.y


TEST_DATA: list[
    str
] = """Sensor at x=2, y=18: closest beacon is at x=-2, y=15
Sensor at x=9, y=16: closest beacon is at x=10, y=16
Sensor at x=13, y=2: closest beacon is at x=15, y=3
Sensor at x=12, y=14: closest beacon is at x=10, y=16
Sensor at x=10, y=20: closest beacon is at x=10, y=16
Sensor at x=14, y=17: closest beacon is at x=10, y=16
Sensor at x=8, y=7: closest beacon is at x=2, y=10
Sensor at x=2, y=0: closest beacon is at x=2, y=10
Sensor at x=0, y=11: closest beacon is at x=2, y=10
Sensor at x=20, y=14: closest beacon is at x=25, y=17
Sensor at x=17, y=20: closest beacon is at x=21, y=22
Sensor at x=16, y=7: closest beacon is at x=15, y=3
Sensor at x=14, y=3: closest beacon is at x=15, y=3
Sensor at x=20, y=1: closest beacon is at x=15, y=3""".split(
    "\n"
)


if __name__ == "__main__":
    testmod()
    z = Zone(stdin.readlines())
    print(z.empty_in_row(2_000_000))
    print(encode(z.find_beacon()))

File: rust/src/bin/aoc_2022_15.rs
// Advent of Code, 2022 day 15

use aoc_rust::stdin_lines;
use aoc_rust::Coord;
use std::collections::HashMap;
use std::collections::HashSet;
use std::iter::zip;

#[derive(Clone, Copy, Debug, PartialEq)]
struct Report {
    sensor: Coord,
    beacon: Coord,
}

fn parse(line: &str) -> Report {
    let parts: Vec<&str> = line.split('=').collect();
    assert_eq!(parts.len(), 5);
    assert_eq!(parts[0], "Sensor at x");
    let numbers: Vec<i32> = parts[1..]
        .iter()
        .map(|s| s.trim_end_matches(|c: char| !c.is_numeric()))
        .map(|s| s.parse().unwrap())
        .collect();
    Report {
        sensor: Coord {
            row: numbers[1],
            col: numbers[0],
        },
        beacon: Coord {
            row: numbers[3],
            col: numbers[2],
        },
    }
}

// Test if the given cell must be empty.
fn must_be_empty(reports: &[Report], test_beacon: Coord) -> bool {
    for r in reports {
        if r.beacon == test_beacon {
            return false;
        }
        let closest_beacon_distance = r.sensor.manhattan(&r.beacon);
        let test_distance = r.sensor.manhattan(&test_beacon);
        if test_distance <= closest_beacon_distance {
            return true;
        }
    }
    false
}

// Return number of positions with given y coordinate that must be empty.
fn empty_in_row(reports: &[Report], y: i32) -> usize {
    let offsets: Vec<Coord> = reports
        .iter()
        .map(|r| Coord {
            row: 0,
            col: r.sensor.manhattan(&r.beacon),
        })
        .collect();
    let x_min: i32 = zip(reports, &offsets)
        .map(|(r, o)| (r.sensor - *o).col)
        .min()
        .unwrap();
    let x_max: i32 = zip(reports, &offsets)
        .map(|(r, o)| (r.sensor + *o).col)
        .max()
        .unwrap();
    (x_min..x_max + 1)
        .filter(|x| must_be_empty(reports, Coord { row: y, col: *x }))
        .count()
}

fn get_constraint(sensor1: Coord, sensor2: Coord, radius1: i32, radius2: i32) -> (bool, i32) {
    assert!(sensor1.row != sensor2.row && sensor1.col != sensor2.col);
    let (s1, s2, r1, r2) = if sensor1.col <= sensor2.col {
        (sensor1, sensor2, radius1, radius2)
    } else {
        (sensor2, sensor1, radius2, radius1)
    };
    assert!(s1.col < s2.col);
    if s1.row < s2.row {
        assert!(s1.col + s1.row + r1 + 1 == s2.col + s2.row - r2 - 1);
        return (true, s1.col + s1.row + r1 + 1);
    }
    assert!(s1.col - s1.row + r1 + 1 == s2.col - s2.row - r2 - 1);
    (false, s1.col - s1.row + r1 + 1)
}

fn find_beacon(reports: &[Report]) -> Coord {
    let sensor_radius: HashMap<Coord, i32> = reports
        .iter()
        .map(|r| (r.sensor, r.sensor.manhattan(&r.beacon)))
        .collect();
    let mut x_plus_y_constraints: HashSet<i32> = HashSet::new();
    let mut x_minus_y_constraints: HashSet<i32> = HashSet::new();
    for report1 in reports {
        for report2 in reports {
            if report1 == report2 {
                continue;
            }
            let s1: Coord = report1.sensor;
            let s2: Coord = report2.sensor;
            let d: i32 = s1.manhattan(&s2);
            let r1: i32 = *sensor_radius.get(&s1).unwrap();
            let r2: i32 = *sensor_radius.get(&s2).unwrap();
            if d == (r1 + 1) + (r2 + 1) {
                let (constraint_sign, constraint_value) = get_constraint(s1, s2, r1, r2);
                if constraint_sign {
                    x_plus_y_constraints.insert(constraint_value);
                } else {
                    x_minus_y_constraints.insert(constraint_value);
                }
            }
        }
    }
    assert!(!x_plus_y_constraints.is_empty() && !x_minus_y_constraints.is_empty());
    let mut solutions: HashSet<Coord> = HashSet::new();
    for p_con in x_plus_y_constraints.iter() {
        for m_con in x_minus_y_constraints.iter() {
            if p_con + m_con % 2 == 1 || p_con - m_con % 2 == 1 {
                continue;
            }
            let col = (p_con + m_con) / 2;
            let row = (p_con - m_con) / 2;
            solutions.insert(Coord { row, col });
        }
    }
    for solution in solutions {
        if reports
            .iter()
            .all(|r| solution.manhattan(&r.sensor) > *sensor_radius.get(&r.sensor).unwrap())
        {
            return solution;
        }
    }
    unreachable!();
}

fn main() {
    let reports: Vec<Report> = stdin_lines().map(|s| parse(&s)).collect();

    println!("{}", empty_in_row(&reports, 2_000_000));
    let beacon = find_beacon(&reports);
    let encoded: i64 = (beacon.col as i64) * 4_000_000 + (beacon.row as i64);
    println!("{}", encoded);
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_parse() {
        let line = "Sensor at x=2, y=18: closest beacon is at x=-2, y=15";
        assert_eq!(
            parse(line),
            Report {
                sensor: Coord { row: 18, col: 2 },
                beacon: Coord { row: 15, col: -2 }
            }
        );
    }
}
